File: apple/RemoteDisplay/Keyboard/Keyboard.swift
//
//  Keyboard.swift
//  RemoteDisplay
//
//  Created by Júlio Flores on 01/09/2026.
//

import Foundation

enum Keyboard {
	enum Action {
		case text(String)
		case backspace
	}

	@MainActor
	protocol Controller: Sendable, AnyObject {
		var canShowKeyboard: Bool { get set }
		func insertText(_ text: String)
		func deleteBackward()
		func toggleKeyboard()
		func setService(_ service: any Service)
		func subscribe(client: any Client)
		func unsubscribe(clientID: UUID)
	}

	@MainActor
	protocol Service: AnyObject {
		func showKeyboard()
		func hideKeyboard()
	}

	@MainActor
	protocol Client: Sendable, Identifiable where ID == UUID {
		func receive(action: Keyboard.Action)
	}
}

File: android/app/src/main/java/dk/cipher/remotedisplay/keyboard/Keyboard.kt
package dk.cipher.remotedisplay.keyboard

object Keyboard {
    sealed interface Action {
        data class Text(val text: String) : Action
        object Backspace : Action
    }

    interface Service {
        fun showKeyboard()
        fun hideKeyboard()
    }

    interface Client {
        fun receive(action: Action)
    }

    interface Controller {
        var canShowKeyboard: Boolean
        fun insertText(text: String)
        fun deleteBackward()
        fun toggleKeyboard()
        fun setService(service: Service)
        fun subscribe(client: Client)
        fun unsubscribe(client: Client)
    }
}